### DISCUSSION TITLE
How to ensure billing with multiple plans?

### DISCUSSION THREAD

**Discussion by bendrobinson1998** [kinngh/shopify-node-express-mongodb-app#127] [Q&A] [answered]
Hey guys - has anyone got any advice on ensuring billing inside this template?

I have managed to add a billing check to the callback route, but I cannot seem to also add that check to the other routes. 

Much appreciated.

**Comment by kinngh** [comment, marked as answer]
Make a middleware that reads it from a KV or from your database

**Reply by bendrobinson1998** [reply]
Thanks man, I cant get any new route handlers in index.js to print a console.log, in particular im trying to billing protect one route, can you spot where im going wrong? The rest of the app is full functional this is my final piece. Appreciate your help: 

`import "@shopify/shopify-api/adapters/node";
import "dotenv/config";
import Express from "express";
import mongoose from "mongoose";
import { resolve } from "path";
import shopify from "../utils/shopify.js";

import sessionHandler from "../utils/sessionHandler.js";
import csp from "./middleware/csp.js";
import setupCheck from "../utils/setupCheck.js";
import {
  customerDataRequest,
  customerRedact,
  shopRedact,
} from "./controllers/gdpr.js";
import applyAuthMiddleware from "./middleware/auth.js";
import isShopActive from "./middleware/isShopActive.js";
import verifyHmac from "./middleware/verifyHmac.js";
import verifyProxy from "./middleware/verifyProxy.js";
import verifyRequest from "./middleware/verifyRequest.js";
import proxyRouter from "./routes/app_proxy/index.js";
import userRoutes from "./routes/index.js";


setupCheck(); // Run a check to ensure everything is setup properly

const PORT = parseInt(process.env.PORT, 10) || 8081;
const isDev = process.env.NODE_ENV === "dev";

// MongoDB Connection
const mongoUrl =
  process.env.MONGO_URL || "***REMOVED***";

mongoose.connect(mongoUrl);

const createServer = async (root = process.cwd()) => {
  const app = Express();


  applyAuthMiddleware(app);
  

// This is the route im trying to trigger a billing check on, but I cant see the log either

  app.get("/debug/submissioncentre", async (req, res, next) => {
    console.log("Sub Triggered");
    try {
      const session = await shopify.session.getCurrentId({
        isOnline: true,
        rawRequest: req,
        rawResponse: res,
      });

      if (!session) {
        return res.redirect(`/api/auth?shop=${req.query.shop}`);
      }

      const hasPayment = await shopify.billing.check({
        session,
        plans: "My plan",
        isTest: true,
      });

      if (hasPayment) {
        // If payment is successful, allow access to the submission centre page
        next();
      } else {
        // If payment is not successful, redirect to the billing request page
        const billingUrl = await shopify.billing.request({
          session,
          plan: "My plan",
          isTest: true,
        });
        return res.redirect(billingUrl);
      }
    } catch (e) {
      console.error(`---> Error at /debug/submissioncentre`, e);
    }
  });

// Incoming webhook requests
  app.post(
    "/api/webhooks/:topic",
    Express.text({ type: "*/*" }),
    async (req, res) => {
      const { topic } = req.params || "";
      const shop = req.headers["x-shopify-shop-domain"] || "";

      try {
        await shopify.webhooks.process({
          rawBody: req.body,
          rawRequest: req,
          rawResponse: res,
        });
        console.log(`--> Processed ${topic} webhook for ${shop}`);
      } catch (e) {
        console.error(
          `---> Error while registering ${topic} webhook for ${shop}`,
          e
        );
        if (!res.headersSent) {
          res.status(500).send(error.message);
        }
      }
    }
  );

  app.use(Express.json());

  app.post("/api/graphql", verifyRequest, async (req, res) => {
    try {
      const sessionId = await shopify.session.getCurrentId({
        isOnline: true,
        rawRequest: req,
        rawResponse: res,
      });
      const session = await sessionHandler.loadSession(sessionId);
      const response = await shopify.clients.graphqlProxy({
        session,
        rawBody: req.body,
      });
      res.status(200).send(response.body);
    } catch (e) {
      console.error(`---> An error occured at GraphQL Proxy`, e);
      res.status(403).send(e);
    }
  });

  app.use(csp);
  app.use(isShopActive);
  // If you're making changes to any of the routes, please make sure to add them in `./client/vite.config.cjs` or it'll not work.
  app.use("/api/apps", verifyRequest, userRoutes); //Verify user route requests
  app.use("/api/proxy_route", verifyProxy, proxyRouter); //MARK:- App Proxy routes

  app.post("/api/gdpr/:topic", verifyHmac, async (req, res) => {
    const { body } = req;
    const { topic } = req.params;
    const shop = req.body.shop_domain;

    console.warn(`--> GDPR request for ${shop} / ${topic} recieved.`);

    let response;
    switch (topic) {
      case "customers_data_request":
        response = await customerDataRequest(topic, shop, body);
        break;
      case "customers_redact":
        response = await customerRedact(topic, shop, body);
        break;
      case "shop_redact":
        response = await shopRedact(topic, shop, body);
        break;
      default:
        console.error(
          "--> Congratulations on breaking the GDPR route! Here's the topic that broke it: ",
          topic
        );
        response = "broken";
        break;
    }

    if (response.success) {
      res.status(200).send();
    } else {
      res.status(403).send("An error occured");
    }
  });

  if (!isDev) {
    const compression = await import("compression").then(
      ({ default: fn }) => fn
    );
    const serveStatic = await import("serve-static").then(
      ({ default: fn }) => fn
    );
    const fs = await import("fs");

    app.use(compression());
    app.use(serveStatic(resolve("dist/client")));
    app.use("/*", (req, res, next) => {
      res
        .status(200)
        .set("Content-Type", "text/html")
        .send(fs.readFileSync(`${root}/dist/client/index.html`));
    });
  }

  return { app };
};

createServer().then(({ app }) => {
  app.listen(PORT, () => {
    console.log(`--> Running on ${PORT}`);
  });
});
`

**Reply by kinngh** [reply]
I'm gonna make an exit on this since I genuinely don't have the time to weed out issues in everyone's code. 

To give you a general idea, subscribe to `APP_SUBSCRIPTIONS_UPDATE` webhook and store the subscriptions in your database. Create a new middleware that reads from the database and either emits `next()` or a `return res.status(...).send(...);` based on what the current subscription plan is.

Also when you paste in huge chunks of code, encapsulate it in \``` and \```, so it's formatted properly. To add proper highlighting, also specify the language. So it should be

\```javascript
// code goes here
\```

This yields in:

```javascript
console.log("So your code looks properly formatted");
console.log("across multiple lines");
```

**Reply by bendrobinson1998** [reply]
Thanks.
